### PR TITLE
BUG: itkDCMTKFileReader Spacing Issue

### PR DESCRIPTION
**Pull request by hossbach-cgm**
… correctly

## PR Checklist
- [x] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [x] Added test (or behavior not changed)
- [x] Updated API documentation (or API not changed)
- [x] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [x] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [x] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

### CODE REVIEW

**Comment by thewtex** [comment]
@issakomi @dclunie @pieper please take a look :pray:

**Comment by jhlegarreta** [comment]
Also, if this is a bug, can a test be added so that we the issue is demonstrated and we ensure that any future change does not cause a regression? Thanks.

**Comment by hossbach-cgm** [comment]
I will certainly clean things up and extend the tests, but for now I find it more important to clarify if this fix is actually a fix, or if I am trying to read something that is actually malformed DICOM.

File: Modules/IO/DCMTK/src/itkDCMTKFileReader.cxx [hunk 1]
```diff
@@ -1250,26 +1253,38 @@ DCMTKFileReader::GetSpacing(double * const spacing) const
...
       }
     }
   }
-  return rval;
+
+  if (_have_val[0] && _have_val[1] && _have_val[2])
+  {
+    return EXIT_SUCCESS;
+  }
+  return EXIT_FAILURE;
```
Comment (issakomi) [inline Modules/IO/DCMTK/src/itkDCMTKFileReader.cxx:1287]: Maybe initialize `_spacing[3]` with `1.0, 1.0, 1.0` and before `return EXIT_FAILURE;` add
```
 spacing[0] = _spacing[0];
 spacing[1] = _spacing[1];
 spacing[2] = _spacing[2];
```
The return value of the function does not seem to be checked, and the double[3] pointer passed to the function is not initialized, so it should be safer. Spacing `0` usually triggers exception later, IMHO, the default is usually `1`.
https://github.com/InsightSoftwareConsortium/ITK/blob/16f42267a00e2c2ae35c189609fad3147ce99e67/Modules/IO/DCMTK/src/itkDCMTKImageIO.cxx#L436

In fact, BTW, I am afraid that support for enhanced multi-frames IODs is rather basic.

**Comment by hossbach-cgm** [comment]
> Also, if this is a bug, can a test be added so that we the issue is demonstrated and we ensure that any future change does not cause a regression? Thanks.

Test added, but I may need help adding the test dicom file to the test data repository.
Also, I tried squashing my commits to be able to clean up the commit message, but I may need some help here...

**Comment by jhlegarreta** [comment]
@hossbach-cgm happy to help.

A few comments:
- Commits have not been squashed. See the commit tab in the PR. I assume no style or changes of different nature are involved, so all should dwell in a single commit. No merge commits, i.e.:
 ```
 $ git checkout master
 $ git fetch upstream
 $ git merge upstream/master
 $ git push origin master
 $ git checkout itkDCMTKFileReaderSpacing
 $ git rebase origin/master
 $ git rebase -i HEAD~N
 (squash)
 ```
assuming you have set your `origin` to your fork, and `upstream` to the ISC repository, and where `N` is the number of commits in this PR.

- Use the appropriate commit message prefix: https://github.com/InsightSoftwareConsortium/ITK/pull/4134/checks?check_run_id=16172192274 and https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#commit-messages.
- The data was added correctly.

Let us know if you need more help with the above.

File: Modules/IO/DCMTK/src/itkDCMTKFileReader.cxx (hunk 1 above)
Comment (pieper) [reply Modules/IO/DCMTK/src/itkDCMTKFileReader.cxx:1296]: > BTW, I am afraid that support for enhanced multi-frame IODs is rather basic.

This is going to be increasingly important.  I've heard that Siemens MR systems will start creating multiframe instances by default soon.

**Comment by pieper** [comment]
My general preference would be to have the core reader code only handle completely standard dicom so that it is readable and maintainable.  Any special code to handle malformed or vendor specific data should be detected and handled in dedicated classes, ideally with diagnostic info available so users know they have odd data and can report the issues to whoever created it.

Also is this behavior going to be followed in GDCM too?  We already have a few cases (e.g. https://github.com/InsightSoftwareConsortium/ITK/issues/4109) where they behave differently.  In Slicer we provide an option to pick either GDCM or DCMTK, but really it's hard for anyone to know which to use for any particular data.

**Comment by thewtex** [comment]
> Test added, but I may need help adding the test dicom file to the test data repository.
Also, I tried squashing my commits to be able to clean up the commit message, but I may need some help here...

@hossbach-cgm I rebased and squashed.

For the testing data, we have a [simplified upload tool](https://content-link.itk.org), which you could try. Or, wrap the file as a `.zip` and attach it to a GitHub comment on this PR.

**Comment by thewtex** [comment]
@hossbach-cgm could you please share a link to the testing data file? Thanks.

**Comment by thewtex** [comment]
@hossbach-cgm thanks!

> My general preference would be to have the core reader code only handle completely standard dicom so that it is readable and maintainable. Any special code to handle malformed or vendor specific data should be detected and handled in dedicated classes, ideally with diagnostic info available so users know they have odd data and can report the issues to whoever created it.

@pieper Good idea. Tracked in #4496.

> Also is this behavior going to be followed in GDCM too?

GDCM already gets the Z-spacing for this dataset correctly. I added another test with this input.

**Review by thewtex — APPROVED** [review]

**Pull request merged**